### PR TITLE
Unicode normalization follow up, adding character navigation and several fixes

### LINKED ISSUE #16616: Unicode normalization: Enable by default for speech and reconsider character navigation
### Is your feature request related to a problem? Please describe.
Recently, Unicode normalization has been added to NVDA, but in #16466, there was some discussion about whether it should be enabled by default.
Furthermore, there was some debate about character navigation, i.e. whether normalization should be enabled by default for character navigation, and/or whether the delayed character description feature could be utilized to some extend.

### Describe the solution you'd like
1. Enable normalization by default for speech output.
2. When enabled, always normalize (i.e. also with character navigation) but prepend the word 'normalized' or something less technical to the character.

### Describe alternatives you've considered
@CyrilleB79 proposed a three choice option in https://github.com/nvaccess/nvda/issues/16466#issuecomment-2132149218:

1. Normalization disabled (Current default)
2. Normalization enabled to read text but disabled to read by character (current behavior of enabled)
3. Normalization enabled for all (not yet implemented)

I'm happy to consider this alternative as well, though it will probably take more effort. I'm now at the point where I think if we add a proper term prepended for character navigation, that would be the best solution.

### Additional context
1. I'm reluctant to take the character descriptions feature into account:

    * Current use of these is for phonetics. I'm inclined to say that adding Unicode character descriptions is a fundamental change to this feature
    * There are no translated names for the several mathematical Unicode chars in CLDR

2. Braille is beyond discussion.

### Discussion guidelines
When attending to the discussion, please avoid repeating standpoints over again that were already stated in https://github.com/nvaccess/nvda/issues/16466. It is clear that we have different opinions and that is okay.

### LINKED ISSUE #16640: Braille unicode normalization unreliable for longer texts with many non-breaking spaces and repeating characters
### Steps to reproduce:
```Python
import textUtils; textUtils.UnicodeNormalizationOffsetConverter('ratie \nen/of rapportage?\xa0\r\n·Structurele taken\xa0of \nscripts\xa0die\xa0periodiek moeten \ndraaien\xa0en\xa0waar\xa0controle\xa0/\xa0logging\xa0omheen \nmoet?\xa0\r\n\xa0\r\nZo ja,\xa0waarmee dient er rekening gehouden te \nworden? Benoem\xa0ook\xa0i')
```

### Actual behavior:
```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "P:\A11y\nvda\source\textUtils.py", line 439, in __init__
    self.computedStrToEncodedOffsets, self.computedEncodedToStrOffsets = self._calculateOffsets()
                                                                         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "P:\A11y\nvda\source\textUtils.py", line 496, in _calculateOffsets
    assert normalizedBuffer == normalizedPart
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

### Expected behavior:
No error

#### NVDA installed/portable/running from source:
Running from source. Installed or portable doesn't assert, but fails later on in the braille module.

#### NVDA version:
Master

### Additional context
I have tracked this down to behavior in difflib:
1. `difflib.ndiff` uses `difflib.Differ` under the hood, which itself uses `difflib.SequenceMatcher`
2. `difflib.SequenceMatcher` has a parameter `autojunk` defaulting to True. It is not possible to override this with neither `ndiff` nor `Differ`
3. When `autojunk` is True, the diff gets scrambled up, treating characters as removed and then inserted further on, effectively messing up our logic to handle the diff

### Possible solution
I'm working on a new implementation based on `SequenceMatcher` directly. The nice thing about `SequenceMatcher` is that it offers:

1. More context: a diff only offers info about insert/delete/equal, whereas `SequenceMatcher` also provides `replacements`. Therefore we can distinguish between modifier reordering and character normalization
2. Direct info about offset ranges.

```Python
import unicodedata, difflib
difflib.SequenceMatcher(None, "Ééĳo", unicodedata.normalize("NFKC", "Ééĳo"), False).get_opcodes()
> Returns: [('replace', 0, 5, 0, 4), ('equal', 5, 6, 4, 5)]
```

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #16622 
fixes #16640 

### Summary of the issue:
It has been discussed that normalization would also be helpful for character navigation. There's also an issue where character descriptions and symbol pronunciation didn't work correctly because normalization took place before symbol processing. Furthermore, for the UnicodeNormalizationOffsetConverter used for braille, it was discovered that diffing didn't turn out to be accurate enough.

### Description of user facing changes
1. When normalization is enabled, there's an extra option `Report normalized when navigating by character` in the speech settings.
2. Added global commands foor speech and braille normalization (without assigned gesture)
3. When normalization is enabled, characters will now always be normalized as well.
4. Braille Unicode normalization is more reliable now.

### Description of development approach
1. Normalization is now always applied to speech, rather than only for object and text info speech. I also changed some helper functions to be able to report normalized when navigating by character.
2. Added a `SuppressUnicodeNormalizationCommand` that allows you to suppress global normalization within a speech sequence. This command is used when creating a spelling sequence, because spelling has its own normalization logic now. It also ensures that when spelling a character or providing a character description (i.e. NVDA+. double press), normalization does not occur. It can even be used to disable character normalization altogether if necessary.
3. Rewrote `UnicodeNormalizationOffsetConverter` to use a new function in NVDAHelper local that uses uniscribe to give a list of offsets for character boundaries. This allows us to split a string into glyphs and then apply normalization to every glyph, which results in more reliable offset calculation and less complex code.

### Testing strategy:
- [x] Test that á and Ĳ are announced in their normalized form when Unicode normalization is on, and that the normalized word is added to the character announcement when enabled. Note that for á, this doesn't always work as several textInfo implementations (such as Mozilla) are slightly broken when navigating across compositions. I must probably report that.
- [x] Test that, even when unicode normalization is off, á is now announced as `a accute` when navigating by character, thereby improving that behavior as well

### Known issues with pull request:
None known

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Added an option to report normalized characters during character navigation.
  - Introduced commands to cycle through Unicode normalization states for braille and speech.
  - Added a checkbox in the speech settings panel for reporting normalized characters.

- **Enhancements**
  - Improved speech synthesis with a new command to suppress Unicode normalization.
  - Enhanced text processing functions to support Unicode normalization.

- **Tests**
  - Added unit tests for notifications and speech normalization, including tests for ligatures and decomposed characters.

- **Documentation**
  - Updated user guide with information on toggling Unicode normalization and reporting normalized characters.

### CODE REVIEW

**Comment by LeonarddeR** [comment]
Thanks for pointing that out. Note that I was aware of the standpoint to delay this pr to 2024.4, but others might not have been.

**Comment by Adriani90** [comment]
@seanbudd  this seems quite stable, which points do you think need still clarification? If you need more community feedback, this needs to be merged into alpha. I don‘t think there is anythink open regarding speech in #16616.  you can also keep this behavior in alpha until you think it is stable enough to bring it into Beta. At least this is how we dealt with new features in the past as well, see e.g. cancellable speech which was enabled only in alpha for a longer period of time until it finally reached the stable version.

**Comment by CyrilleB79** [comment]
@seanbudd would you accept milestone 2024.3 for this PR if we change default value to disabled (see https://github.com/nvaccess/nvda/issues/16624#issuecomment-2136630649)? This would at least avoid to ship 2024.3 with a buggy feature, closing #16624.

A subsequent PR for 2024.4 could then switch the default value to enabled.

**Comment by seanbudd** [comment]
@CyrilleB79 - yes as requested in https://github.com/nvaccess/nvda/issues/16624#issuecomment-2136614990, however @LeonarddeR seems to suggest in https://github.com/nvaccess/nvda/issues/16624#issuecomment-2136630649 that this is not possible for whatever reason

**Comment by LeonarddeR** [comment]
@seanbudd  I must have misunderstood you then. I thought you were asking in https://github.com/nvaccess/nvda/issues/16624#issuecomment-2136614990 whether it would be possible to fix #16624 without changing character navigation behavior. That would be very difficult. If you'd agree changing this pr to disable normalization again by default but yet change character navigation to also normalize, that is certainly possibly and I can change the pr as requested by @CyrilleB79.

**Comment by seanbudd** [comment]
That would be great yes. The normalization fix can go into 2024.3, but we don't want to turn it on by default until it has had some wider testing in a release

**Comment by LeonarddeR** [comment]
@seanbudd I think this is ready for a review. Would you be able to trigger a signed try build for broader testing?

**Comment by seanbudd** [comment]
A try build should be built soon: https://ci.appveyor.com/project/NVAccess/nvda/builds/49964259, thanks for waiting

File: user_docs/en/userGuide.md [hunk 1]
```diff
@@ -2151,6 +2160,8 @@ When this option is enabled, unicode normalization is performed on the text that
 This is beneficial when coming across characters in braille that are unknown in a particular braille table and which have a compatible alternative, like the bold and italic characters commonly used on social media.
 Other benefits of unicode normalization are explained in greater detail in the [section for the equivalent speech setting](#SpeechUnicodeNormalization).
 
+To toggle Unicode normalization from anywhere, please assign a custom gesture using the [Input Gestures dialog](#InputGestures).
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply user_docs/en/userGuide.md:2163]: I think you are mistaking our markdown syntax. Header anchors are specified through tags like this: `#Header {#Anchor}`

**Comment by Adriani90** [comment]
Note that a trybuild is not appropriate for wide community testing, because this will need to be distrobuted manually in all communities which is an overload in this case. I really hope you reconsider enabling this by default in alpha so people can test and report directly without having to discover this setting manually.

**Comment by Adriani90** [comment]
Also @LeonarddeR if this should still be a feature flag, which is by the way not really needed in my opinion, then this settings should be moved to advanced settings category. Otherwise it would be nice if you convert the combo boxes to checkboxes in order to be in line with NVDA GUI conventions.

**Comment by Adriani90** [comment]
Could you move them then please to advanced settings in a dedicated grouping? This is the common understanding when having feature flags.Von meinem iPhone gesendetAm 06.06.2024 um 10:37 schrieb Leonard de Ruijter ***@***.***>:﻿

Also @LeonarddeR if this should still be a feature flag, which is by the way not really needed in my opinion, then this settings should be moved to advanced settings category. Otherwise it would be nice if you convert the combo boxes to checkboxes in order to be in line with NVDA GUI conventions.

The Show selection option in braille settings is also a feature flag.
I'm pretty sure @seanbudd yet prefers this to stay feature flags.

**Comment by XLTechie** [comment]
@seanbudd, I also had the understanding that feature flag combos were (1) for pre-release or wide-scale testing features that we weren't sure we were going to keep, or weren't sure should have the default they do during testing; and
(2) were expected to only appear in Advanced, where users did not have to be taught the concept, or how to handle them.

A regular checkbox should be sufficient for a mature feature that is being released as ready for general use, and the most appropriate default should be selected and used by the time it arrives in a normal panel.
That probably means they can only be converted during a release with a config upgrade, but I had assumed that was the expected usecase.

I can not think of any other software that uses this kind of setting in their stable interface.

It is true that, for example, VoiceOver, uses a similar concept in its equivalent to config profiles, to indicate "follow whatever the default mode for this setting is", but that is not really the same.

File: source/textUtils.py [hunk 2]
```diff
@@ -438,9 +439,29 @@ def __init__(self, text: str, normalizationForm: str = DEFAULT_UNICODE_NORMALIZA
...
+		but it doesn't allow us to control the autojunk parameter of SequenceMatcher.
+		Diffing with autojunk on sometimes breaks the logic to calculate offsets for unicode normalization.
+		This context manager, when active, ensures that ndiff uses a monkey patched SequenceMatcher
+		with autojunk disabled.
+		"""
+		try:
+			SequenceMatcher.autojunk = property(lambda sm: False, lambda sm, v: None)
+			yield
+		finally:
+			del SequenceMatcher.autojunk
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/textUtils.py:455]: This suggestion doesn't make sense, as autojunk is a property that is only set on instances, i.e. it isn't available on the class by default. This code ensures that we add a temporary property to the class so setting a truety value on the instance is ignored.

File: tests/unit/test_speech.py [hunk 3]
```diff
@@ -194,19 +194,35 @@ def test_capNotificationsWithPlaceHolderBefore(self):
...
+			' normalized'
+		])
+		output = _getSpellingCharAddCapNotification(
+			speakCharAs='A',
+			sayCapForCapitals=False,
+			capPitchChange=0,
+			beepForCapitals=False,
+			reportNormalized=True,
+		)
+		self.assertEqual(repr(list(output)), expected)
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply tests/unit/test_speech.py:209]: No. The cases where it is False are already covered by other cases.

**Comment by LeonarddeR** [comment]
Okay, we really need to take a different approach. Diffing is not reliable enough.
Let's try uniscribe to the rescue!

**Comment by LeonarddeR** [comment]
@coderabbitai review
could you have a look into textUtils.cpp? May be we can work around some duplicated code here and there?

**Comment by LeonarddeR** [comment]
@coderabbitai I'd rather fix duplication between calculateCharacterBoundaries and _calculateUniscribeOffsets. Note that the former swaps some c style code for c++, but a major part of both is very similar.

**Comment by LeonarddeR** [comment]
@coderabbitai I didn't like your suggestions much. I refactored it in the last commit.

**Comment by seanbudd** [comment]
@LeonarddeR -is this ready for re-review?

I'd encourage any change of policy/usage of feature flags to be moved to a separate discussion.
Right now they are encouraged for general settings where appropriate, and are used throughout the braille settings, as well as in paragraph nav and other general settings.

**Comment by LeonarddeR** [comment]
@seanbudd Yes, I think the unit tests, including some edge cases, seem to prove that this works as advertised.

**Review by Qchristensen — APPROVED** [review]
UserGuide looks good.

File: source/gui/settingsDialogs.py [hunk 4]
```diff
@@ -1597,9 +1597,24 @@ def makeSettings(self, settingsSizer):
 			wxCtrlClass=nvdaControls.FeatureFlagCombo,
 			keyPath=["speech", "unicodeNormalization"],
 			conf=config.conf,
+			onChoiceEventHandler=self._onUnicodeNormalizationChange,
 		)
 		self.bindHelpEvent("SpeechUnicodeNormalization", self.unicodeNormalizationCombo)
 
+		# Translators: This is the label for a checkbox in the
+		# speech settings panel.
+		reportNormalizedForCharacterNavigationText = _("Report &Normalized when navigating by character")
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:1606]:
```suggestion
		reportNormalizedForCharacterNavigationText = _("Report '&Normalized' when navigating by character")
```

File: source/speech/speech.py [hunk 5]
```diff
@@ -160,10 +161,17 @@ def isBlank(text):
 RE_CONVERT_WHITESPACE = re.compile("[\0\r\n]")
 
 
-def processText(locale: str, text: str, symbolLevel: characterProcessing.SymbolLevel) -> str:
+def processText(
+		locale: str,
+		text: str,
+		symbolLevel: characterProcessing.SymbolLevel,
+		normalize: bool = False
+) -> str:
```
Comment (seanbudd) [inline source/speech/speech.py:169]: can you add a docstring, particularly for the behaviour of normalise

File: source/speech/speech.py [hunk 6]
```diff
@@ -331,14 +339,17 @@ def _getSpellingCharAddCapNotification(
...
 	"""This function produces a speech sequence containing a character to be spelt as well as commands
-	to indicate that this character is uppercase if applicable.
-	@param speakCharAs: The character as it will be spoken by the synthesizer.
-	@param sayCapForCapitals: indicates if 'cap' should be reported along with the currently spelt character.
-	@param capPitchChange: pitch offset to apply while spelling the currently spelt character.
-	@param beepForCapitals: indicates if a cap notification beep should be produced while spelling the currently
-	spellt character.
+	to indicate that this character is uppercase and/or normalized, if applicable.
+	:param speakCharAs: The character as it will be spoken by the synthesizer.
+	:param sayCapForCapitals: indicates if 'cap' should be reported along with the currently speled character.
```
Comment (seanbudd) [inline source/speech/speech.py:347]:
```suggestion
	:param sayCapForCapitals: indicates if 'cap' should be reported along with the currently spelled character.
```

File: source/speech/speech.py [hunk 7]
```diff
@@ -331,14 +339,17 @@ def _getSpellingCharAddCapNotification(
...
+	:param capPitchChange: pitch offset to apply while spelling the currently speled character.
```
Comment (seanbudd) [inline source/speech/speech.py:348]:
```suggestion
	:param capPitchChange: pitch offset to apply while spelling the currently spelled character.
```

File: source/speech/speech.py [hunk 8]
```diff
@@ -331,14 +339,17 @@ def _getSpellingCharAddCapNotification(
...
+	:param beepForCapitals: indicates if a cap notification beep should be produced while spelling the currently
+	spelled character.
+	:param reportNormalized: indicates if 'normalized' should be reported
+	along with the currently speled character.
```
Comment (seanbudd) [inline source/speech/speech.py:352]:
```suggestion
	along with the currently spelled character.
```

File: source/speech/speech.py [hunk 9]
```diff
@@ -369,6 +390,8 @@ def _getSpellingSpeechWithoutCharMode(
 		capPitchChange: int,
 		beepForCapitals: bool,
 		fallbackToCharIfNoDescription: bool = True,
+		unicodeNormalization: bool = False,
+		reportNormalizedForCharacterNavigation: bool = False,
```
Comment (seanbudd) [inline source/speech/speech.py:394]: please add docstring entries for new parameters

File: source/textUtils/__init__.py [hunk 10]
```diff
@@ -429,122 +429,39 @@ class UnicodeNormalizationOffsetConverter(OffsetConverter):
...
-			originResult,
-			normalizedResult
-		))
+				computedStrToEncodedOffsets.extend(normOffset for origChar in origPart)
+				computedEncodedToStrOffsets.extend(origOffset for normChar in normPart)
+			origOffset += len(origPart)
+			normOffset += len(normPart)
+		self.encoded = normalized
+
+	def _processReordered(self, a: str, b: str):
```
Comment (seanbudd) [inline source/textUtils/__init__.py:459]: please add a return type to this

File: source/textUtils/uniscribe.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+	Contrary to just iterating over a string, this respects surrogate pairs, decomposite characters, etc.
+	"""
+	import NVDAHelper  # Import late to avoid circular import.
+
+	if not NVDAHelper.localLib:
+		raise RuntimeError("NVDAHelper not initialized")
+	if not text:
+		return
+	# uniscribe does some strange things
+	# when you give it a string  with not more than two alphanumeric chars in a row.
```
Comment (seanbudd) [inline source/textUtils/uniscribe.py:24]:
```suggestion
	# when you give it a string with not more than two alphanumeric chars in a row.
```

File: source/textUtils/uniscribe.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+	Splits a given string into real visible characters (or glyphs), thereby respecting character boundaries.
+	Contrary to just iterating over a string, this respects surrogate pairs, decomposite characters, etc.
+	"""
+	import NVDAHelper  # Import late to avoid circular import.
+
+	if not NVDAHelper.localLib:
+		raise RuntimeError("NVDAHelper not initialized")
+	if not text:
+		return
+	# uniscribe does some strange things
```
Comment (seanbudd) [inline source/textUtils/uniscribe.py:23]: is there an issue reference for this?

File: source/textUtils/uniscribe.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+	# when you give it a string  with not more than two alphanumeric chars in a row.
+	# Inject two alphanumeric characters at the end to fix this
```
Comment (seanbudd) [inline source/textUtils/uniscribe.py:25]: should we be handling the end of this buffer somewhere when returning data? does this affect results

**Comment by Adriani90** [comment]
@seanbudd 
> I'd encourage any change of policy/usage of feature flags to be moved to a separate discussion.
Right now they are encouraged for general settings where appropriate, and are used throughout the braille settings, as well as in paragraph nav and other general settings.

This is because people didn't pay attention to the actual purpose of feature flags when reviewing the coresponding PRs. I think it makes sense to clearly not encourage using these feature flags outside of the advanced settings panel anymore, and state the actuual purpose in the policy docs.
Many users find this feature flag approach outside of the advanced settings as quite bad UX.

File: source/speech/speech.py (hunk 5 above)
Comment (LeonarddeR) [reply source/speech/speech.py:169]: Sure, will do

File: source/speech/speech.py (hunk 9 above)
Comment (LeonarddeR) [reply source/speech/speech.py:394]: I'll update the doc string.

File: source/textUtils/uniscribe.py (hunk 12 above)
Comment (LeonarddeR) [reply source/textUtils/uniscribe.py:23]: I'm afraid not. I copied this from `OffsetsTextInfo._calculateUniscribeOffsets` where a similar workaround was applied for similar code.

File: source/textUtils/uniscribe.py (hunk 13 above)
Comment (LeonarddeR) [reply source/textUtils/uniscribe.py:25]: The two characters are filtered out of the result.
The only impact this had, is that even when the `text` parameter was an empty string, the NVDAHelper function would still be called with just "xx". I fixed that by explicitly checking for an empty string.

**Comment by seanbudd** [comment]
is this ready for re-review?

**Comment by LeonarddeR** [comment]
Yes, sorry. Forgot to mark it as ready

**Review by seanbudd — APPROVED** [review]
